### DISCUSSION TITLE
WhatChanged for single entity

### DISCUSSION THREAD

**Discussion by stevehansen** [ravendb/ravendb#17324] [Q&A] [answered]
We are trying to do some simple generic IAudit logic, so before going to the database we want to set some properties on the changed object.

We are using the IDocumentStore.OnBeforeStore hook, which is invoked per entity but we want to get some information so we are using the session.Advanced.WhatChanged(), the problem is that this method will look for all changed entities.

When a small set of entities are loaded in the session the performance isn't that bad, but when we have a few hundred or thousand (which still seem small) the performance gets really bad. In our case we got the issue when using the RavenMigrations package to run some migrations, we loaded the whole collection to do the modification (about 1500 documents), the code also had to load some extra related entities and the memory usage spiked to almost 10GB during development.

![image](https://github.com/ravendb/ravendb/assets/2143488/ffc138c5-8579-4afa-9600-53bf02d13c61)

The code of course does some more but the root of the problem is the many WhatChanged() calls, that's why we are wondering if there is another option to get the result from it, ie the DocumentChanges for a single entity.

**Comment by ayende** [comment]
The API for comparing changes is here:

https://github.com/ravendb/ravendb/blob/7a58173cb9a469af65b729f926acb5397030e225/src/Raven.Client/Json/BlittableOperation.cs#L29

However, you need the _previous_ version of the instance for this to work.

That is here:
https://github.com/ravendb/ravendb/blob/cf8f473ba47f754ac1a9228e92d455d21ccfb8a3/src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs#L374

Which is a private method.

**Reply by stevehansen** [reply]
similar to the methods above the GetDocumentInfo method would you be willing to accept a PR that adds the following method?

```c#
/// <summary>
/// Gets the document changes for the specified instance.
/// Throws an exception if the instance is not tracked by the session.
/// </summary>
/// <param name="instance">The instance.</param>
/// <returns></returns>
public DocumentsChanges[] GetWhatChangedFor<T>(T instance)
```

**Reply by ppekrol** [reply]
Yes, but let's make it:
```csharp
public DocumentsChanges[] WhatChangedFor(object entity)
```

**Reply by stevehansen** [reply]
I added the generic because the GetDocumentInfo also expects the generic argument, GetChangedVectorFor/GetLastModifiedFor pass the generic as-is to the GetDocumentInfo so not sure what happens if we pass object directly

**Reply by ppekrol** [reply]
It will be fine, what you actually want is to +/- copy `HasChanged` method but pass the changes there to fill

**Comment by ThierryRosseau** [comment]
https://issues.hibernatingrhinos.com/issue/RavenDB-21648/Add-WhatChangedFor

**Comment by ppekrol** [comment, marked as answer]
Merged: https://github.com/ravendb/ravendb/pull/17620
